nbconvert/exporters/slides.py

```diff
@@ -1,28 +1,78 @@
 """HTML slide show Exporter class"""
 
-#-----------------------------------------------------------------------------
-# Copyright (c) 2013, the IPython Development Team.
-#
+# Copyright (c) Jupyter Development Team.
 # Distributed under the terms of the Modified BSD License.
-#
-# The full license is in the file COPYING.txt, distributed with this software.
-#-----------------------------------------------------------------------------
 
-#-----------------------------------------------------------------------------
-# Imports
-#-----------------------------------------------------------------------------
+from copy import deepcopy
 
-from nbconvert import preprocessors
 from traitlets.config import Config
+from traitlets import Unicode
 
 from .html import HTMLExporter
 
-#-----------------------------------------------------------------------------
-# Classes
-#-----------------------------------------------------------------------------
+def prepare(nb):
+    """Add some convenenience metadata on cells for the slide template.
+    """
+    nb = deepcopy(nb)
+
+    for cell in nb.cells:
+        # Make sure every cell has a slide_type
+        cell.metadata.slide_type = cell.metadata.get('slideshow', {}).get('slide_type', '-')
+
+    # Find the first visible cell
+    for index, cell in enumerate(nb.cells):
+        if cell.metadata.slide_type not in {'notes', 'skip'}:
+            cell.metadata.slide_type = 'slide'
+            cell.metadata.slide_start = True
+            cell.metadata.subslide_start = True
+            first_slide_ix = index
+            break
+    else:
+        raise ValueError("All cells are hidden, cannot create slideshow")
+
+    in_fragment = False
+
+    for index, cell in enumerate(nb.cells[first_slide_ix+1:],
+                                 start=(first_slide_ix+1)):
+
+        previous_cell = nb.cells[index - 1]
+
+        # Get the slide type. If type is start, subslide, or slide,
+        # end the last subslide/slide.
+        if cell.metadata.slide_type == 'slide':
+            previous_cell.metadata.slide_end = True
+            cell.metadata.slide_start = True
+        if cell.metadata.slide_type in {'subslide', 'slide'}:
```

Comment:
we should make a comment here to help understand people that we defined our slides as 
```
<section>
  <section>
  ...
  </section>
<section>```

otherwise is really confusing why we check the slide_type of this set (`{'subslide', 'slide'}`).